src/Text/Pandoc/Writers/LaTeX.hs

```diff
@@ -1486,30 +1489,25 @@ citationsToBiblatex (c:cs)
                     SuppressAuthor -> "\\autocites*"
                     AuthorInText   -> "\\textcites"
                     NormalCitation -> "\\autocites"
+      
+      groups <- mapM citeArgumentsList (reverse (foldl' grouper [] (c:cs)))
 
-      let groups = map correctOrder $ reverse $ foldl' grouper [([[]], [])] (c:cs)
-      docGroups <- mapM citeArgsList groups      
-
-      return $ text cmd <> (mconcat docGroups)
+      return $ text cmd <> (mconcat groups)
 
   where grouper prev cit = grouper' (citationPrefix cit)
                (citationSuffix cit) (citationId cit) 
             where grouper' [] sfx cid 
                      | null sfx = addToGroup id 
-                     | otherwise = addToGroup ((:) sfx) 
+                     | otherwise = addToGroup (\(CiteGroup pfx _ ids) -> 
+                            CiteGroup pfx sfx ids) 
                      where addToGroup fn = case prev of  
-                             (((pfx:pfxs), ids):rest) 
-                                 | null pfxs -> (fn [pfx], cid:ids):rest
-                                 | otherwise -> (fn $ pfxs, [cid]):prev
-                             (([], ids):rest) -> (fn [[]], cid:ids):rest
-                             [] -> (fn [[]], [cid]):prev
-                  grouper' pfx [] cid = ([pfx], [cid]):prev
-                  grouper' pfx sfx cid = ([sfx, pfx], [cid]):prev 
-
-        correctOrder (pfxs, ids) = (reverse pfxs, reverse ids)
-        citeArgsList ((p:s:_), k) = citeArgumentsList p s k
-        citeArgsList ((p:_), k) = citeArgumentsList p [] k
-        citeArgsList ([], k) = citeArgumentsList [] [] k
+                             ((CiteGroup pfx oSfx ids):rest) 
+                                 | null oSfx ->
+                                     (fn (CiteGroup pfx oSfx (cid:ids))):rest
+                                 | otherwise -> (fn (CiteGroup pfx [] [cid])):prev
+                             [] -> (fn (CiteGroup [] [] [cid])):prev
+                  grouper' pfx [] cid = (CiteGroup pfx [] [cid]):prev
+                  grouper' pfx sfx cid = (CiteGroup pfx sfx [cid]):prev 
 
```

Comment:
This is so much clearer now!  The only improvement I'd suggest now would be flattening out the logic in `grouper`.  It's awfully hard to read the code with three levels of nested `where`s - `grouper`, then `grouper'`, then `addToGroup`.  I think I'd understand better if it were flattened out, ideally into one level.  Maybe there's a reason for not doing this, but it's what I'd suggest for code readability.  E.g. my first reaction to line 1501 was "wait, is the previous suffix just being thrown away?"  But then I discovered that the function is never applied to a CiteGroup with non-null suffix. It would be better if the logic were more perspicuous here.